File: backend/app/services/customer_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException

from app.models.customer import Customer
from app.schemas.customer import CustomerCreate, CustomerUpdate


def create_customer(db: Session, customer: CustomerCreate):
    existing = (
        db.query(Customer)
        .filter(Customer.phone == customer.phone)
        .first()
    )

    if existing:
        raise HTTPException(
            status_code=400,
            detail="Customer already exists"
        )

    db_customer = Customer(
        name=customer.name,
        phone=customer.phone,
        email=customer.email,
        address=customer.address,
    )

    db.add(db_customer)
    db.commit()
    db.refresh(db_customer)

    return db_customer


def get_all_customers(
    db: Session,
    skip: int = 0,
    limit: int = 10,
):
    return (
        db.query(Customer)
        .offset(skip)
        .limit(limit)
        .all()
    )


def get_customer_by_id(
    db: Session,
    customer_id: int,
):
    return (
        db.query(Customer)
        .filter(Customer.id == customer_id)
        .first()
    )


def update_customer(
    db: Session,
    customer_id: int,
    customer: CustomerUpdate,
):
    db_customer = get_customer_by_id(
        db,
        customer_id,
    )

    if db_customer is None:
        return None

    db_customer.name = customer.name
    db_customer.phone = customer.phone
    db_customer.email = customer.email
    db_customer.address = customer.address

    db.commit()
    db.refresh(db_customer)

    return db_customer


def delete_customer(
    db: Session,
    customer_id: int,
):
    customer = get_customer_by_id(
        db,
        customer_id,
    )

    if customer is None:
        return None

    db.delete(customer)
    db.commit()

    return customer


def search_customers(
    db: Session,
    name: str,
):
    return (
        db.query(Customer)
        .filter(Customer.name.ilike(f"%{name}%"))
        .all()
    )

File: frontend/src/services/customer.service.ts
import api from "./api";

export async function getCustomers() {
  const response = await api.get("/customers");
  return response.data;
}

export async function getCustomer(id: number) {
  const response = await api.get(`/customers/${id}`);
  return response.data;
}

export async function createCustomer(customer: {
  name: string;
  phone: string;
  email: string;
  address: string;
}) {
  const response = await api.post("/customers", customer);
  return response.data;
}

export async function updateCustomer(
  id: number,
  customer: {
    name: string;
    phone: string;
    email: string;
    address: string;
  }
) {
  const response = await api.put(`/customers/${id}`, customer);
  return response.data;
}

export async function deleteCustomer(id: number) {
  const response = await api.delete(`/customers/${id}`);
  return response.data;
}

export async function searchCustomers(name: string) {
  const response = await api.get(`/customers/search?name=${name}`);
  return response.data;
}